src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs

```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
         }
 
         /// <summary>
-        /// Returns the first circular dependency found
+        /// Returns the first circular dependency found. Please sort solution topologically first to improve performance.
         /// </summary>
         public static IEnumerable<ResolverPackage> FindCircularDependency(IEnumerable<ResolverPackage> solution)
         {
-            // check each package to see if it is part of a loop, sort by id to keep the result deterministic
-            foreach (var package in solution.OrderBy(package => package.Id, StringComparer.OrdinalIgnoreCase))
-            {
-                var result = FindCircularDependency(package, Enumerable.Empty<string>(), solution);
+            // to keep track of visited packages to avoid processing them again
+            var visitedPackages = new HashSet<ResolverPackage>();
+
+            var packageLookUp = solution.ToDictionary(p => p.Id, StringComparer.OrdinalIgnoreCase);
 
-                if (result.Any())
+            // check each package to see if it is part of a loop, sort by id to keep the result deterministic
+            foreach (var package in solution)
+            {                
+                var result = FindCircularDependency(package, packageLookUp, visitedPackages);
+                if (result?.Any() == true)
```

Comment:
this only returns the first circular reference (which I think is intentional), but that means the comment and method name should reflect it (so perhaps `FindFirstCircularDependency`), and the comment should say check to find the first package that has a circular dependency.